File: Album/src/main/java/com/hqq/album/activity/AlbumPreviewActivity.java
package com.hqq.album.activity;

import android.app.Activity;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.PagerSnapHelper;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.hqq.album.Adapter.PreviewAdapter;
import com.hqq.album.AppManager;
import com.hqq.album.R;
import com.hqq.album.activity.base.BaseActivity;
import com.hqq.album.common.FunctionKey;
import com.hqq.album.common.FunctionOptions;
import com.hqq.album.common.SelectOptions;
import com.hqq.album.dialog.OptAnimationLoader;
import com.hqq.album.entity.LocalMedia;
import com.hqq.album.customize.FilterImageView;

import java.util.List;

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.album.activity
 * @FileName :   AlbumPreviewV2Activity
 * @Date : 2019/9/16 0016  下午 8:42
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public class AlbumPreviewActivity extends BaseActivity implements View.OnClickListener {


    private RecyclerView mRcAlbumList;
    private ConstraintLayout mAlbumTitleBar;
    private FilterImageView mAlbumBack;
    private TextView mAlbumTitle;
    private LinearLayout mLlCheck;
    private ImageView mTvCheck;
    private LinearLayout mLlComplete;
    private TextView mAlbumFinish;
    List<LocalMedia> mLocalMediaList;
    int mPosition = 0;
    PreviewAdapter mPreviewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        this.requestWindowFeature(Window.FEATURE_NO_TITLE);
//        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);

        setContentView(R.layout.activity_album_preview_v2);
        initView();
        mRcAlbumList.setLayoutManager(new LinearLayoutManager(this, RecyclerView.HORIZONTAL, false));
        new PagerSnapHelper().attachToRecyclerView(mRcAlbumList);
        mRcAlbumList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    int first = ((LinearLayoutManager) recyclerView.getLayoutManager()).findFirstVisibleItemPosition();
                    int last = ((LinearLayoutManager) recyclerView.getLayoutManager()).findLastVisibleItemPosition();
                    if (first == last && mPosition != last) {
                        mPosition = last;
                        updateSelectMenu();
                    }
                }

            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
            }
        });
        mPosition = getIntent().getIntExtra(FunctionKey.KEY_POSITION, 1) - 1;
        mRcAlbumList.scrollToPosition(mPosition);
        updateSelectMenu();

    }

    @Override
    public void onClick(View v) {
        int i = v.getId();
        if (i == R.id.album_back) {
            AppManager.getAppManager().finishActivity();
        } else if (i == R.id.album_finish) {

            setResult(Activity.RESULT_OK);
            AppManager.getAppManager().finishAllActivityAndCallBack();

        } else if (i == R.id.ll_check) {
            if (mTvCheck.isSelected()) {
                mTvCheck.setSelected(false);
                for (LocalMedia media : SelectOptions.getInstance().getSelectLocalMedia()) {
                    if (media.getPath().equals(mLocalMediaList.get(mPosition).getPath())) {
                        SelectOptions.getInstance().getSelectLocalMedia().remove(media);
                        break;
                    }
                }
            } else {
                if (SelectOptions.getInstance().getSelectLocalMedia().size() < FunctionOptions.getInstance().getMaxSelectNum()) {
                    SelectOptions.getInstance().getSelectLocalMedia().add(mLocalMediaList.get(mPosition));
                    mTvCheck.setSelected(true);
                    mTvCheck.startAnimation(OptAnimationLoader.loadAnimation(mContext, R.anim.modal_in));
                } else {
                    Toast.makeText(mContext, mContext.getString(R.string.picture_message_max_num, FunctionOptions.getInstance().getMaxSelectNum() + ""), Toast.LENGTH_LONG).show();
                }
            }
            updateSelectMenu();

        }
    }


    private void initView() {
        mRcAlbumList = (RecyclerView) findViewById(R.id.rc_album_list);
        mAlbumTitleBar = (ConstraintLayout) findViewById(R.id.album_title_bar);
        mAlbumBack = (FilterImageView) findViewById(R.id.album_back);
        mAlbumTitle = (TextView) findViewById(R.id.album_title);
        mLlCheck = (LinearLayout) findViewById(R.id.ll_check);
        mTvCheck = (ImageView) findViewById(R.id.tv_check);
        mLlComplete = (LinearLayout) findViewById(R.id.ll_complete);
        mAlbumFinish = (TextView) findViewById(R.id.album_finish);

        //  RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) mAlbumTitleBar.getLayoutParams();
        //    layoutParams.setMargins(0, AlbumUtils.getStatusBarHeight(this), 0, 0);
        //   mAlbumTitleBar.setLayoutParams(layoutParams);

        mAlbumBack.setOnClickListener(this);
        mLlCheck.setOnClickListener(this);
        mAlbumFinish.setOnClickListener(this);

        mLocalMediaList = SelectOptions.getInstance().getFolderLocalMedia();

        mAlbumTitle.setText(getIntent().getIntExtra(FunctionKey.KEY_POSITION, 1) + "/" + mLocalMediaList.size());
        mPreviewAdapter = new PreviewAdapter(this, mLocalMediaList);
        mRcAlbumList.setItemViewCacheSize(3);
        mRcAlbumList.setAdapter(mPreviewAdapter);

    }

    private void updateSelectMenu() {
        if ((SelectOptions.getInstance().getSelectLocalMedia().size() > 0)) {
            mAlbumFinish.setVisibility(View.VISIBLE);
            mAlbumFinish.setText("完成(" + SelectOptions.getInstance().getSelectLocalMedia().size() + "/" + FunctionOptions.getInstance().getMaxSelectNum() + ")");
        } else {
            mAlbumFinish.setVisibility(View.GONE);
        }
        mAlbumTitle.setText(mPosition + 1 + "/" + mLocalMediaList.size());
        mTvCheck.setSelected(isSelected(mLocalMediaList.get(mPosition)));

    }

    /**
     * 当前图片是否选中
     *
     * @param image
     * @return
     */
    public boolean isSelected(LocalMedia image) {
        for (LocalMedia media : SelectOptions.getInstance().getSelectLocalMedia()) {
            if (media.getPath().equals(image.getPath())) {
                return true;
            }
        }
        return false;
    }


}

File: Album/src/main/java/com/hqq/album/activity/AlbumPreviewActivity.kt
package com.hqq.album.activity

import android.content.DialogInterface
import android.graphics.Bitmap
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.hqq.album.Adapter.PreviewAdapter
import com.hqq.album.AppManager
import com.hqq.album.R
import com.hqq.album.annotation.LocalMediaType.VALUE_TYPE_IMAGE
import com.hqq.album.annotation.LocalMediaType.VALUE_URL_IMAGE
import com.hqq.album.common.FunctionKey
import com.hqq.album.common.FunctionOptions
import com.hqq.album.common.SelectOptions
import com.hqq.album.databinding.ActivityAlbumPreviewV2Binding
import com.hqq.album.dialog.OptAnimationLoader
import com.hqq.album.entity.LocalMedia
import com.hqq.core.glide.ImageLoadUtils
import com.hqq.core.ui.base.BaseActivity
import com.hqq.core.ui.base.BaseViewBindingActivity
import com.hqq.core.ui.base.open
import com.hqq.core.ui.dialog.SelectDialog
import com.hqq.core.utils.ToastUtils
import com.hqq.core.utils.file.FileUtils
import com.hqq.core.utils.file.SaveBitmapUtils

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.album.activity
 * @FileName :   AlbumPreviewV2Activity
 * @Date : 2019/9/16 0016  下午 8:42
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
class AlbumPreviewActivity : BaseAlbumActivity<ActivityAlbumPreviewV2Binding>(), View.OnClickListener {
    companion object {
        /**
         * @param activity BaseActivity
         * @param list ArrayList<String>  url 访问的预览
         */
        fun openPreview(activity: BaseActivity, list: ArrayList<String>) {
            val loc = ArrayList<LocalMedia>()
            for (s in list) {
                loc.add(LocalMedia(s, 0L, 0, VALUE_URL_IMAGE))
            }
            SelectOptions.instance.mFolderLocalMedia = loc

            activity.open(AlbumPreviewActivity::class.java, Bundle().apply {
                putBoolean("isShowSelect", false)
                putBoolean("isSavaBitmap", true)
            })
        }
    }


    var mLocalMediaList: List<LocalMedia>? = null
    var mPosition = 0
    var mPreviewAdapter: PreviewAdapter = PreviewAdapter()

    /**
     *  是否 可以勾选
     */
    var _isShowSelect = true

    /**
     *  是否可以保存图片
     */
    var _isSavaBitmap = false

    override fun initConfig() {
        super.initConfig()
        rootViewImpl.iToolBarBuilder.showLine = false
        rootViewImpl.iToolBarBuilder.showToolBar = false
    }

    override fun initView() {
        _isShowSelect = getBoolean("isShowSelect", true)
        _isSavaBitmap = getBoolean("isSavaBitmap", false)
        if (_isShowSelect) {
            binding.llCheck.visibility = View.VISIBLE
        } else {
            binding.llCheck.visibility = View.GONE
        }

        var currPosition = intent.getIntExtra(FunctionKey.KEY_POSITION, 1)
        mPosition = currPosition - 1
        mLocalMediaList = SelectOptions.instance.mFolderLocalMedia
        mPreviewAdapter.setNewInstance(mLocalMediaList?.toMutableList())
        initSaveBitmap()



        binding.albumBack.setOnClickListener(this)
        binding.llCheck.setOnClickListener(this)
        binding.albumFinish.setOnClickListener(this)
        binding.albumTitle.text = currPosition.toString() + "/" + (mLocalMediaList?.size ?: 0)


        binding.rcAlbumList.setItemViewCacheSize(3)
        binding.rcAlbumList.adapter = mPreviewAdapter

        binding.rcAlbumList.layoutManager = LinearLayoutManager(this, RecyclerView.HORIZONTAL, false)
        PagerSnapHelper().attachToRecyclerView(binding.rcAlbumList)
        binding.rcAlbumList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    val first = (recyclerView.layoutManager as LinearLayoutManager?)!!.findFirstVisibleItemPosition()
                    val last = (recyclerView.layoutManager as LinearLayoutManager?)!!.findLastVisibleItemPosition()
                    if (first == last && mPosition != last) {
                        mPosition = last
                        updateSelectMenu()
                    }
                }
            }
        })
        binding.rcAlbumList.scrollToPosition(mPosition)
        updateSelectMenu()

    }

    private fun initSaveBitmap() {
        if (_isSavaBitmap) {
            mPreviewAdapter.call = {
                SelectDialog.Builder()
                    .setContent("是否保存图片")
                    .setPositiveButton("确定", object : DialogInterface.OnClickListener {
                        override fun onClick(dialog: DialogInterface?, which: Int) {
                            ImageLoadUtils.getBitmapByFail(activity, it.path, object :
                                ImageLoadUtils.GlideLoadBitmapCallback {
                                override fun getBitmapCallback(bitmap: Bitmap) {
                                    SaveBitmapUtils.saveBitmap2Pictures(bitmap, "", FileUtils.getDefFileName(".png"))
                                    ToastUtils.showToast("保存成功")
                                }

                                override fun onLoadFailed() {
                                }
                            })
                            dialog?.dismiss()
                        }
                    })
                    .create()
                    .show(supportFragmentManager)
            }
        }
    }

    override fun onClick(v: View) {
        val i = v.id
        if (i == R.id.album_back) {
            onBackPressed()
        } else if (i == R.id.album_finish) {
            setResult(RESULT_OK)
            AppManager.appManager?.finishAllActivityAndCallBack()
        } else if (i == R.id.ll_check) {
            if (binding.tvCheck.isSelected) {
                binding.tvCheck.isSelected = false
                for (media in SelectOptions.instance.selectLocalMedia) {
                    if (media.path == mLocalMediaList!![mPosition].path) {
                        SelectOptions.instance.selectLocalMedia.remove(media)
                        break
                    }
                }
            } else {
                if (SelectOptions.instance.selectLocalMedia.size < FunctionOptions.instance.maxSelectNum) {
                    SelectOptions.instance.selectLocalMedia.add(mLocalMediaList!![mPosition])
                    binding.tvCheck.isSelected = true
                    binding.tvCheck.startAnimation(activity.let { OptAnimationLoader.loadAnimation(it, R.anim.modal_in) })
                } else {
                    Toast.makeText(activity, activity?.getString(R.string.picture_message_max_num, FunctionOptions.instance.maxSelectNum.toString() + ""), Toast.LENGTH_LONG)
                        .show()
                }
            }
            updateSelectMenu()
        }
    }

    private fun updateSelectMenu() {
        if (SelectOptions.instance.selectLocalMedia.size > 0) {
            binding.albumFinish.visibility = View.VISIBLE
            binding.albumFinish.text =
                "完成(" + SelectOptions.instance.selectLocalMedia.size + "/" + FunctionOptions.instance.maxSelectNum + ")"
        } else {
            binding.albumFinish.visibility = View.GONE
        }
        binding.albumTitle.text = (mPosition + 1).toString() + "/" + mLocalMediaList!!.size
        binding.tvCheck.isSelected = isSelected(mLocalMediaList!![mPosition])
    }

    /**
     * 当前图片是否选中
     *
     * @param image
     * @return
     */
    fun isSelected(image: LocalMedia): Boolean {
        for (media in SelectOptions.instance.selectLocalMedia) {
            if (media.path == image.path) {
                return true
            }
        }
        return false
    }
}